terminus-ui/popover/src/popover-trigger.directive.ts

```diff
@@ -0,0 +1,302 @@
+  }
+  public get position(): TsPopoverPosition {
+    return this._position;
+  }
+  public _position = TsPopoverPositions.Bottom;
+
+  /**
+   * TsPopoverComponent provided as an input
+   */
+  @Input()
+  public popover!: TsPopoverComponent;
+
+  /**
+   * Emit when create popover.
+   */
+  @Output()
+  public readonly popoverOnCreate = new EventEmitter<TsPopoverComponent>();
+
+  /**
+   * Emit when popover is shown.
+   */
+  @Output()
+  public readonly popoverOnShown = new EventEmitter<TsPopoverComponent>();
+
+  /**
+   * Emit when popover is hidden.
+   */
+  @Output()
+  public readonly popoverOnHidden = new EventEmitter<TsPopoverComponent>();
+
+  /**
+   * Emit when popover is updated.
+   */
+  @Output()
+  public readonly popoverOnUpdate = new EventEmitter<TsPopoverComponent>();
+
+  /**
+   * Register the click to decide whether it's inside the element or out.
+   * @param targetElement
+   */
+  @HostListener('document:click', ['$event.target'])
+  public onClick(targetElement) {
+    const clickedInside = this.popover.popoverViewRef.nativeElement.contains(targetElement);
+    if (!clickedInside && this.hideOnBlur) {
+      this.hide();
+    }
+  }
+
+  constructor(
+    private viewContainerRef: ViewContainerRef,
+    private changeDetectorRef: ChangeDetectorRef,
+    private elementRef: ElementRef,
+    private ngZone: NgZone,
+    private documentService: TsDocumentService,
+  ) {
+    /**
+     * Listen to `keydown` events outside the zone so that change detection is not run every
+     * time a key is pressed. Re-enter the zone only if the `ESC` key is pressed
+     */
+    this.ngZone.runOutsideAngular(() => {
+      (fromEvent<KeyboardEvent>(this.elementRef.nativeElement, 'keydown'))
+        .pipe(
+          filter(event => event.code === KEYS.ESCAPE.code),
+          untilComponentDestroyed(this)
+        ).subscribe(event => this.ngZone.run(() => {
+          this.hide();
+          event.stopPropagation();
+          event.preventDefault();
+        }));
+    });
+    this.document = this.documentService.document;
+
+  }
+
+  /**
+   * When directive initiated, it assigns a reference to popover component and also register on click event.
+   */
+  public ngOnInit(): void {
+    this.popover.referenceObject = this.viewContainerRef.element.nativeElement;
+    this.setContentProperties(this.popover);
+
+    (fromEvent<MouseEvent>(this.elementRef.nativeElement, 'click'))
+      .pipe(
+        untilComponentDestroyed(this)
+      ).subscribe(event => {
+        this.toggle();
+        event.stopPropagation();
+        event.preventDefault();
+      });
+  }
+
+  /**
+   * Called after the default change detector has completed checking all content
+   */
+  public ngAfterContentChecked(): void {
+    if (this.defaultOpened) {
+      this.show();
+    }
+    if (this.popover) {
+      this.popover.myId = this.id;
+    }
+  }
+
+  /**
+   * Called on every change happens
+   * @param changes - SimpleChange
```

Comment:
I think we can better describe what is happening here. Currently we are just describing what the lifecycle event does.

Also, missing a line break before the params.